Intermittent TestFileSystem failures

Here is a sample stack trace:
```
18:51:25 00:12.264 org.h2.test.unit.TestFileSystem Expected: -1792929026 actual: -369484912Exception in thread 
org.h2.test.unit.TestFileSystem$13:1867" java.lang.AssertionError: Expected: -1792929026 actual: -369484912
	at org.h2.test.TestBase.fail(TestBase.java:327)
	at org.h2.test.TestBase.assertEquals(TestBase.java:499)
	at org.h2.test.unit.TestFileSystem$13.call(TestFileSystem.java:788)
	at org.h2.util.Task.run(Task.java:44)
	at java.lang.Thread.run(Thread.java:748)```
```

It looks like FileEncrypt (and XTS object within it in particular) is not thread-safe, but I can't figure out why.
Synchronizing XTS.encrypt() and XTS.decrypt() seems to fix the problem though.

> I don't see failures with changes from that commit on JDK 8 and 11 on Linux x86_64.

It's all about timing, so it could be number of cores (4 hyper) or disk (SSD), but still weird, cause mine is failing left and right:
```
08:24:33 64:22:15.860 org.h2.test.unit.TestFileSystem Done pass #0
08:24:35 64:22:17.741 org.h2.test.unit.TestFileSystem Done pass #1
08:24:35 64:22:18.322 org.h2.test.unit.TestFileSystem Expected: -1934797313 actual: 1110650121 message: encrypt:0007:./data/test/fs at 2
Exception in thread "org.h2.test.unit.TestFileSystem$13:2" java.lang.AssertionError: Expected: -1934797313 actual: 1110650121 message: encrypt:0007:./data/test/fs at 2
	at org.h2.test.TestBase.fail(TestBase.java:327)
	at org.h2.test.TestBase.assertEquals(TestBase.java:486)
	at org.h2.test.unit.TestFileSystem$13.call(TestFileSystem.java:791)
	at org.h2.util.Task.run(Task.java:44)
	at java.base/java.lang.Thread.run(Thread.java:834)
08:24:37 64:22:19.516 org.h2.test.unit.TestFileSystem Done pass #2
08:24:38 64:22:21.314 org.h2.test.unit.TestFileSystem Done pass #3
08:24:40 64:22:23.080 org.h2.test.unit.TestFileSystem Done pass #4
08:24:42 64:22:24.827 org.h2.test.unit.TestFileSystem Done pass #5
08:24:44 64:22:26.578 org.h2.test.unit.TestFileSystem Done pass #6
08:24:45 64:22:28.364 org.h2.test.unit.TestFileSystem Done pass #7
08:24:46 64:22:28.890 org.h2.test.unit.TestFileSystem Expected: 508436724 actual: 279155947 message: encrypt:0007:./data/test/fs at 3
Exception in thread "org.h2.test.unit.TestFileSystem$13:8" java.lang.AssertionError: Expected: 508436724 actual: 279155947 message: encrypt:0007:./data/test/fs at 3
	at org.h2.test.TestBase.fail(TestBase.java:327)
	at org.h2.test.TestBase.assertEquals(TestBase.java:486)
	at org.h2.test.unit.TestFileSystem$13.call(TestFileSystem.java:791)
	at org.h2.util.Task.run(Task.java:44)
	at java.base/java.lang.Thread.run(Thread.java:834)
08:24:47 64:22:30.082 org.h2.test.unit.TestFileSystem Done pass #8
08:24:49 64:22:31.807 org.h2.test.unit.TestFileSystem Done pass #9
08:24:50 64:22:33.102 org.h2.test.unit.TestFileSystem Expected: 467697586 actual: 190494464 message: encrypt:0007:./data/test/fs at 6
Exception in thread "org.h2.test.unit.TestFileSystem$13:10" java.lang.AssertionError: Expected: 467697586 actual: 190494464 message: encrypt:0007:./data/test/fs at 6
	at org.h2.test.TestBase.fail(TestBase.java:327)
	at org.h2.test.TestBase.assertEquals(TestBase.java:486)
	at org.h2.test.unit.TestFileSystem$13.call(TestFileSystem.java:791)
	at org.h2.util.Task.run(Task.java:44)
	at java.base/java.lang.Thread.run(Thread.java:834)
08:24:51 64:22:33.643 org.h2.test.unit.TestFileSystem Done pass #10
08:24:52 64:22:35.383 org.h2.test.unit.TestFileSystem Done pass #11
08:24:54 64:22:37.162 org.h2.test.unit.TestFileSystem Done pass #12
08:24:56 64:22:38.953 org.h2.test.unit.TestFileSystem Done pass #13
08:24:58 64:22:40.706 org.h2.test.unit.TestFileSystem Done pass #14
08:24:59 64:22:42.417 org.h2.test.unit.TestFileSystem Done pass #15
08:25:00 64:22:42.762 org.h2.test.unit.TestFileSystem Expected: -1119952315 actual: 993585977 message: encrypt:0007:./data/test/fs at 4
Exception in thread "org.h2.test.unit.TestFileSystem$13:16" java.lang.AssertionError: Expected: -1119952315 actual: 993585977 message: encrypt:0007:./data/test/fs at 4
	at org.h2.test.TestBase.fail(TestBase.java:327)
	at org.h2.test.TestBase.assertEquals(TestBase.java:486)
	at org.h2.test.unit.TestFileSystem$13.call(TestFileSystem.java:791)
	at org.h2.util.Task.run(Task.java:44)
	at java.base/java.lang.Thread.run(Thread.java:834)
08:25:01 64:22:44.107 org.h2.test.unit.TestFileSystem Done pass #16
08:25:03 64:22:45.819 org.h2.test.unit.TestFileSystem Done pass #17
08:25:05 64:22:47.549 org.h2.test.unit.TestFileSystem Done pass #18
08:25:06 64:22:49.281 org.h2.test.unit.TestFileSystem Done pass #19
08:25:08 64:22:51.040 org.h2.test.unit.TestFileSystem Done pass #20
08:25:10 64:22:52.739 org.h2.test.unit.TestFileSystem Done pass #21
08:25:12 64:22:54.576 org.h2.test.unit.TestFileSystem Done pass #22
08:25:13 64:22:56.340 org.h2.test.unit.TestFileSystem Done pass #23
08:25:14 64:22:56.601 org.h2.test.unit.TestFileSystem Expected: 1633795756 actual: 1453184918 message: encrypt:0007:./data/test/fs at 9
Exception in thread "org.h2.test.unit.TestFileSystem$13:24" java.lang.AssertionError: Expected: 1633795756 actual: 1453184918 message: encrypt:0007:./data/test/fs at 9
	at org.h2.test.TestBase.fail(TestBase.java:327)
	at org.h2.test.TestBase.assertEquals(TestBase.java:486)
	at org.h2.test.unit.TestFileSystem$13.call(TestFileSystem.java:791)
	at org.h2.util.Task.run(Task.java:44)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

--- a/h2/src/docsrc/html/changelog.html
+++ b/h2/src/docsrc/html/changelog.html
@@ -21,6 +21,8 @@ Change Log
 
 <h2>Next Version (unreleased)</h2>
 <ul>
+<li>Issue #2757: Intermittent TestFileSystem failures
+</li>
 <li>Issue #2758: Issues with sequences
 </li>
 <li>PR #2756: Prevent DROP NOT NULL for identity columns

--- a/h2/src/docsrc/html/features.html
+++ b/h2/src/docsrc/html/features.html
@@ -1100,10 +1100,11 @@ The port can be set manually using <code>AUTO_SERVER_PORT=9090</code>.
 
 <h2 id="page_size">Page Size</h2>
 <p>
-The page size for new databases is 2 KB (2048), unless the page size is set
+The page size for new databases is 4 KiB (4096 bytes), unless the page size is set
 explicitly in the database URL using <code>PAGE_SIZE=</code> when
 the database is created. The page size of existing databases can not be changed,
 so this property needs to be set when the database is created.
+The page size of encrypted databases must be a multiple of 4096 (4096, 8192, &#8230;).
 </p>
 
 <h2 id="trace_options">Using the Trace Options</h2>

--- a/h2/src/main/org/h2/engine/Database.java
+++ b/h2/src/main/org/h2/engine/Database.java
@@ -68,6 +68,7 @@ import org.h2.store.InDoubtTransaction;
 import org.h2.store.LobStorageFrontend;
 import org.h2.store.LobStorageInterface;
 import org.h2.store.fs.FileUtils;
+import org.h2.store.fs.encrypt.FileEncrypt;
 import org.h2.table.Column;
 import org.h2.table.IndexColumn;
 import org.h2.table.Table;
@@ -260,7 +261,10 @@ public class Database implements DataHandler, CastDataProvider {
         this.autoServerPort = ci.getProperty("AUTO_SERVER_PORT", 0);
         int defaultCacheSize = Utils.scaleForAvailableMemory(Constants.CACHE_SIZE_DEFAULT);
         this.cacheSize = ci.getProperty("CACHE_SIZE", defaultCacheSize);
-        this.pageSize = ci.getProperty("PAGE_SIZE", Constants.DEFAULT_PAGE_SIZE);
+        pageSize = ci.getProperty("PAGE_SIZE", Constants.DEFAULT_PAGE_SIZE);
+        if (cipher != null && pageSize % FileEncrypt.BLOCK_SIZE != 0) {
+            throw DbException.getUnsupportedException("CIPHER && PAGE_SIZE=" + pageSize);
+        }
         if ("r".equals(accessModeData)) {
             readOnly = true;
         }

--- a/h2/src/main/org/h2/store/fs/encrypt/FileEncrypt.java
+++ b/h2/src/main/org/h2/store/fs/encrypt/FileEncrypt.java
@@ -60,7 +60,7 @@ public class FileEncrypt extends FileBaseDefault {
 
     private final String name;
 
-    private XTS xts;
+    private volatile XTS xts;
 
     private byte[] encryptionKey;
 
@@ -73,9 +73,19 @@ public class FileEncrypt extends FileBaseDefault {
         this.encryptionKey = encryptionKey;
     }
 
-    private void init() throws IOException {
+    private XTS init() throws IOException {
+        // Keep this method small to allow inlining
+        XTS xts = this.xts;
+        if (xts == null) {
+            xts = createXTS();
+        }
+        return xts;
+    }
+
+    private synchronized XTS createXTS() throws IOException {
+        XTS xts = this.xts;
         if (xts != null) {
-            return;
+            return xts;
         }
         this.size = base.size() - HEADER_LENGTH;
         boolean newFile = size < 0;
@@ -94,10 +104,9 @@ public class FileEncrypt extends FileBaseDefault {
             }
         }
         AES cipher = new AES();
-        cipher.setKey(SHA256.getPBKDF2(
-                encryptionKey, salt, HASH_ITERATIONS, 16));
+        cipher.setKey(SHA256.getPBKDF2(encryptionKey, salt, HASH_ITERATIONS, 16));
         encryptionKey = null;
-        xts = new XTS(cipher);
+        return this.xts = new XTS(cipher);
     }
 
     @Override
@@ -111,7 +120,7 @@ public class FileEncrypt extends FileBaseDefault {
         if (len == 0) {
             return 0;
         }
-        init();
+        XTS xts = init();
         len = (int) Math.min(len, size - position);
         if (position >= size) {
             return -1;
@@ -126,19 +135,18 @@ public class FileEncrypt extends FileBaseDefault {
             int offset = (int) (position - p);
             int l = (len + offset + BLOCK_SIZE - 1) / BLOCK_SIZE * BLOCK_SIZE;
             ByteBuffer temp = ByteBuffer.allocate(l);
-            readInternal(temp, p, l);
+            readInternal(temp, p, l, xts);
             temp.flip();
             temp.limit(offset + len);
             temp.position(offset);
             dst.put(temp);
             return len;
         }
-        readInternal(dst, position, len);
+        readInternal(dst, position, len, xts);
         return len;
     }
 
-    private void readInternal(ByteBuffer dst, long position, int len)
-            throws IOException {
+    private void readInternal(ByteBuffer dst, long position, int len, XTS xts) throws IOException {
         int x = dst.position();
         readFully(base, position + HEADER_LENGTH, dst);
         long block = position / BLOCK_SIZE;
@@ -162,7 +170,7 @@ public class FileEncrypt extends FileBaseDefault {
 
     @Override
     public int write(ByteBuffer src, long position) throws IOException {
-        init();
+        XTS xts = init();
         int len = src.remaining();
         if ((position & BLOCK_SIZE_MASK) != 0 ||
                 (len & BLOCK_SIZE_MASK) != 0) {
@@ -175,7 +183,7 @@ public class FileEncrypt extends FileBaseDefault {
             int available = (int) (size - p + BLOCK_SIZE - 1) / BLOCK_SIZE * BLOCK_SIZE;
             int readLen = Math.min(l, available);
             if (readLen > 0) {
-                readInternal(temp, p, readLen);
+                readInternal(temp, p, readLen, xts);
                 temp.rewind();
             }
             temp.limit(offset + len);
@@ -183,7 +191,7 @@ public class FileEncrypt extends FileBaseDefault {
             temp.put(src);
             temp.limit(l);
             temp.rewind();
-            writeInternal(temp, p, l);
+            writeInternal(temp, p, l, xts);
             long p2 = position + len;
             size = Math.max(size, p2);
             int plus = (int) (size & BLOCK_SIZE_MASK);
@@ -193,14 +201,13 @@ public class FileEncrypt extends FileBaseDefault {
             }
             return len;
         }
-        writeInternal(src, position, len);
+        writeInternal(src, position, len, xts);
         long p2 = position + len;
         size = Math.max(size, p2);
         return len;
     }
 
-    private void writeInternal(ByteBuffer src, long position, int len)
-            throws IOException {
+    private void writeInternal(ByteBuffer src, long position, int len, XTS xts) throws IOException {
         ByteBuffer crypt = ByteBuffer.allocate(len);
         crypt.put(src);
         crypt.flip();

--- a/h2/src/main/org/h2/store/fs/encrypt/FileEncrypt.java
+++ b/h2/src/main/org/h2/store/fs/encrypt/FileEncrypt.java
@@ -127,8 +127,7 @@ public class FileEncrypt extends FileBaseDefault {
         } else if (position < 0) {
             throw new IllegalArgumentException("pos: " + position);
         }
-        if ((position & BLOCK_SIZE_MASK) != 0 ||
-                (len & BLOCK_SIZE_MASK) != 0) {
+        if ((position & BLOCK_SIZE_MASK) != 0 || (len & BLOCK_SIZE_MASK) != 0) {
             // either the position or the len is unaligned:
             // read aligned, and then truncate
             long p = position / BLOCK_SIZE * BLOCK_SIZE;
@@ -136,9 +135,7 @@ public class FileEncrypt extends FileBaseDefault {
             int l = (len + offset + BLOCK_SIZE - 1) / BLOCK_SIZE * BLOCK_SIZE;
             ByteBuffer temp = ByteBuffer.allocate(l);
             readInternal(temp, p, l, xts);
-            temp.flip();
-            temp.limit(offset + len);
-            temp.position(offset);
+            temp.flip().limit(offset + len).position(offset);
             dst.put(temp);
             return len;
         }
@@ -157,8 +154,7 @@ public class FileEncrypt extends FileBaseDefault {
         }
     }
 
-    private static void readFully(FileChannel file, long pos, ByteBuffer dst)
-            throws IOException {
+    private static void readFully(FileChannel file, long pos, ByteBuffer dst) throws IOException {
         do {
             int len = file.read(dst, pos);
             if (len < 0) {
@@ -172,8 +168,7 @@ public class FileEncrypt extends FileBaseDefault {
     public int write(ByteBuffer src, long position) throws IOException {
         XTS xts = init();
         int len = src.remaining();
-        if ((position & BLOCK_SIZE_MASK) != 0 ||
-                (len & BLOCK_SIZE_MASK) != 0) {
+        if ((position & BLOCK_SIZE_MASK) != 0 || (len & BLOCK_SIZE_MASK) != 0) {
             // either the position or the len is unaligned:
             // read aligned, and then truncate
             long p = position / BLOCK_SIZE * BLOCK_SIZE;
@@ -186,11 +181,8 @@ public class FileEncrypt extends FileBaseDefault {
                 readInternal(temp, p, readLen, xts);
                 temp.rewind();
             }
-            temp.limit(offset + len);
-            temp.position(offset);
-            temp.put(src);
-            temp.limit(l);
-            temp.rewind();
+            temp.limit(offset + len).position(offset);
+            temp.put(src).limit(l).rewind();
             writeInternal(temp, p, l, xts);
             long p2 = position + len;
             size = Math.max(size, p2);
@@ -208,8 +200,7 @@ public class FileEncrypt extends FileBaseDefault {
     }
 
     private void writeInternal(ByteBuffer src, long position, int len, XTS xts) throws IOException {
-        ByteBuffer crypt = ByteBuffer.allocate(len);
-        crypt.put(src);
+        ByteBuffer crypt = ByteBuffer.allocate(len).put(src);
         crypt.flip();
         long block = position / BLOCK_SIZE;
         int x = 0, l = len;
@@ -221,12 +212,9 @@ public class FileEncrypt extends FileBaseDefault {
         writeFully(base, position + HEADER_LENGTH, crypt);
     }
 
-    private static void writeFully(FileChannel file, long pos,
-            ByteBuffer src) throws IOException {
-        int off = 0;
+    private static void writeFully(FileChannel file, long pos, ByteBuffer src) throws IOException {
         do {
-            int len = file.write(src, pos + off);
-            off += len;
+            pos += file.write(src, pos);
         } while (src.remaining() > 0);
     }
 
@@ -260,8 +248,7 @@ public class FileEncrypt extends FileBaseDefault {
     }
 
     @Override
-    public FileLock tryLock(long position, long size, boolean shared)
-            throws IOException {
+    public FileLock tryLock(long position, long size, boolean shared) throws IOException {
         return base.tryLock(position, size, shared);
     }
 

--- a/h2/src/main/org/h2/store/fs/encrypt/FileEncrypt.java
+++ b/h2/src/main/org/h2/store/fs/encrypt/FileEncrypt.java
@@ -24,7 +24,7 @@ public class FileEncrypt extends FileBaseDefault {
     /**
      * The block size.
      */
-    static final int BLOCK_SIZE = 4096;
+    public static final int BLOCK_SIZE = 4096;
 
     /**
      * The block size bit mask.

--- a/h2/src/test/org/h2/test/db/TestEncryptedDb.java
+++ b/h2/src/test/org/h2/test/db/TestEncryptedDb.java
@@ -39,6 +39,8 @@ public class TestEncryptedDb extends TestDb {
     @Override
     public void test() throws SQLException {
         deleteDb("encrypted");
+        assertThrows(ErrorCode.FEATURE_NOT_SUPPORTED_1, this).
+                getConnection("encrypted;CIPHER=AES;PAGE_SIZE=2048", "sa", "1234 1234");
         try (Connection conn = getConnection("encrypted;CIPHER=AES", "sa", "123 123")) {
             Statement stat = conn.createStatement();
             stat.execute("CREATE TABLE TEST(ID INT)");

--- a/h2/src/test/org/h2/test/unit/TestFileSystem.java
+++ b/h2/src/test/org/h2/test/unit/TestFileSystem.java
@@ -21,6 +21,7 @@ import java.sql.SQLException;
 import java.sql.Statement;
 import java.util.List;
 import java.util.Random;
+import java.util.concurrent.atomic.AtomicIntegerArray;
 import java.util.zip.ZipEntry;
 import java.util.zip.ZipOutputStream;
 import org.h2.dev.fs.FilePathZip2;
@@ -770,6 +771,7 @@ public class TestFileSystem extends TestBase {
         final FileChannel f = FileUtils.open(s, "rw");
         final int size = getSize(10, 50);
         f.write(ByteBuffer.allocate(size * 64 *  1024));
+        AtomicIntegerArray locks = new AtomicIntegerArray(size);
         Random random = new Random(1);
         System.gc();
         Task task = new Task() {
@@ -779,7 +781,13 @@ public class TestFileSystem extends TestBase {
                 while (!stop) {
                     for (int pos = 0; pos < size; pos++) {
                         byteBuff.clear();
-                        f.read(byteBuff, pos * 64 * 1024);
+                        while (!locks.compareAndSet(pos, 0, 1)) {
+                        }
+                        try {
+                            f.read(byteBuff, pos * 64 * 1024);
+                        } finally {
+                            locks.set(pos, 0);
+                        }
                         byteBuff.position(0);
                         int x = byteBuff.getInt();
                         int y = byteBuff.getInt();
@@ -800,11 +808,23 @@ public class TestFileSystem extends TestBase {
                 byteBuff.putInt(i);
                 byteBuff.flip();
                 int pos = random.nextInt(size);
-                f.write(byteBuff, pos * 64 * 1024);
+                while (!locks.compareAndSet(pos, 0, 1)) {
+                }
+                try {
+                    f.write(byteBuff, pos * 64 * 1024);
+                } finally {
+                    locks.set(pos, 0);
+                }
                 data[pos] = i;
                 pos = random.nextInt(size);
                 byteBuff.clear();
-                f.read(byteBuff, pos * 64 * 1024);
+                while (!locks.compareAndSet(pos, 0, 1)) {
+                }
+                try {
+                    f.read(byteBuff, pos * 64 * 1024);
+                } finally {
+                    locks.set(pos, 0);
+                }
                 byteBuff.limit(16);
                 byteBuff.position(0);
                 int x = byteBuff.getInt();

--- a/h2/src/test/org/h2/test/unit/TestFileSystem.java
+++ b/h2/src/test/org/h2/test/unit/TestFileSystem.java
@@ -772,6 +772,7 @@ public class TestFileSystem extends TestBase {
         final int size = getSize(10, 50);
         f.write(ByteBuffer.allocate(size * 64 *  1024));
         AtomicIntegerArray locks = new AtomicIntegerArray(size);
+        AtomicIntegerArray expected = new AtomicIntegerArray(size);
         Random random = new Random(1);
         System.gc();
         Task task = new Task() {
@@ -781,9 +782,11 @@ public class TestFileSystem extends TestBase {
                 while (!stop) {
                     for (int pos = 0; pos < size; pos++) {
                         byteBuff.clear();
+                        int e;
                         while (!locks.compareAndSet(pos, 0, 1)) {
                         }
                         try {
+                            e = expected.get(pos);
                             f.read(byteBuff, pos * 64 * 1024);
                         } finally {
                             locks.set(pos, 0);
@@ -791,14 +794,14 @@ public class TestFileSystem extends TestBase {
                         byteBuff.position(0);
                         int x = byteBuff.getInt();
                         int y = byteBuff.getInt();
-                        assertEquals(x, y);
+                        assertEquals(e, x);
+                        assertEquals(e, y);
                         Thread.yield();
                     }
                 }
             }
         };
         task.execute();
-        int[] data = new int[size];
         try {
             ByteBuffer byteBuff = ByteBuffer.allocate(16);
             int operations = 10000;
@@ -812,15 +815,17 @@ public class TestFileSystem extends TestBase {
                 }
                 try {
                     f.write(byteBuff, pos * 64 * 1024);
+                    expected.set(pos, i);
                 } finally {
                     locks.set(pos, 0);
                 }
-                data[pos] = i;
                 pos = random.nextInt(size);
                 byteBuff.clear();
+                int e;
                 while (!locks.compareAndSet(pos, 0, 1)) {
                 }
                 try {
+                    e = expected.get(pos);
                     f.read(byteBuff, pos * 64 * 1024);
                 } finally {
                     locks.set(pos, 0);
@@ -829,8 +834,8 @@ public class TestFileSystem extends TestBase {
                 byteBuff.position(0);
                 int x = byteBuff.getInt();
                 int y = byteBuff.getInt();
-                assertEquals(x, y);
-                assertEquals(data[pos], x);
+                assertEquals(e, x);
+                assertEquals(e, y);
             }
         } catch (Throwable e) {
             e.printStackTrace();

